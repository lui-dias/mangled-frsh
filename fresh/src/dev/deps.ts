// std
export {
  basename,
  dirname,
  extname,
  fromFileUrl,
  join,
  posix,
  relative,
  resolve,
  SEP,
  toFileUrl,
} from "https://deno.land/std@0.193.0/path/mod.ts";
export { DAY, WEEK } from "https://deno.land/std@0.193.0/datetime/constants.ts";
export * as colors from "https://deno.land/std@0.193.0/fmt/colors.ts";
export {
  walk,
  type WalkEntry,
  WalkError,
} from "https://deno.land/std@0.193.0/fs/walk.ts";
export { parse } from "https://deno.land/std@0.193.0/flags/mod.ts";
export { gte } from "https://deno.land/std@0.193.0/semver/mod.ts";
export { existsSync } from "https://deno.land/std@0.193.0/fs/mod.ts";
export * as semver from "https://deno.land/std@0.195.0/semver/mod.ts";
export * as JSONC from "https://deno.land/std@0.195.0/jsonc/mod.ts";
export * as fs from "https://deno.land/std@0.195.0/fs/mod.ts";

// ts-morph
export { Node, Project } from "https://deno.land/x/ts_morph@17.0.1/mod.ts";
