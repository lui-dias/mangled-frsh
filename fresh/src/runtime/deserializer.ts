// Run `deno run -A npm:esbuild --minify src/runtime/deserializer.ts` to minify
// this file. It is embedded into src/server/deserializer_code.ts.

export const KEY = "_f";

interface Signal<T> {
  peek(): T;
  value: T;
}

function b64decode(b64: string): Uint8Array {
  const binString = atob(b64);
  const size = binString.length;
  const bytes = new Uint8Array(size);
  for (let i = 0; i < size; i++) {
    bytes[i] = binString.charCodeAt(i);
  }
  return bytes;
}

export function deserialize(
  str: string,
  signal?: <T>(a: T) => Signal<T>,
): unknown {
  function reviver(this: unknown, _key: string, value: unknown): unknown {
    if (typeof value === "object" && value && KEY in value) {
      // deno-lint-ignore no-explicit-any
      const v: any = value;
      if (v[KEY] === "s") {
        return signal!(v.v);
      }
      if (v[KEY] === "b") {
        return BigInt(v.d);
      }
      if (v[KEY] === "u8a") {
        return b64decode(v.d);
      }
      if (v[KEY] === "l") {
        const val = v.v;
        val[KEY] = v.k;
        return val;
      }
      throw new Error(`Unknown key: ${v[KEY]}`);
    }
    return value;
  }

  const { v, r } = JSON.parse(str, reviver);
  const references = (r ?? []) as [string[], ...string[][]][];
  for (const [targetPath, ...refPaths] of references) {
    const target = targetPath.reduce((o, k) => k === null ? o : o[k], v);
    for (const refPath of refPaths) {
      if (refPath.length === 0) throw new Error("Invalid reference");
      // set the reference to the target object
      const parent = refPath.slice(0, -1).reduce(
        (o, k) => k === null ? o : o[k],
        v,
      );
      parent[refPath[refPath.length - 1]!] = target;
    }
  }
  return v;
}
