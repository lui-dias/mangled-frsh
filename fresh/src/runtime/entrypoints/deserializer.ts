export { deserialize } from "../deserializer.ts";
