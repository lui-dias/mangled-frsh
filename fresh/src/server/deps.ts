// -- std --
export {
  dirname,
  extname,
  fromFileUrl,
  isAbsolute,
  join,
  toFileUrl,
} from "https://deno.land/std@0.193.0/path/mod.ts";
export { walk } from "https://deno.land/std@0.193.0/fs/walk.ts";
export * as colors from "https://deno.land/std@0.193.0/fmt/colors.ts";
export {
  type Handler as ServeHandler,
  serve,
} from "https://deno.land/std@0.193.0/http/server.ts";
export { Status } from "https://deno.land/std@0.193.0/http/http_status.ts";
export {
  typeByExtension,
} from "https://deno.land/std@0.193.0/media_types/mod.ts";
export { toHashString } from "https://deno.land/std@0.193.0/crypto/to_hash_string.ts";
export { escape } from "https://deno.land/std@0.193.0/regexp/escape.ts";
export * as JSONC from "https://deno.land/std@0.193.0/jsonc/mod.ts";
