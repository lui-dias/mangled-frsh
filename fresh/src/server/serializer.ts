/**
 * This module contains a serializer for island props. The serializer is capable
 * of serializing the following:
 *
 * - `null`
 * - `boolean`
 * - `number`
 * - `bigint`
 * - `string`
 * - `array`
 * - `object` (no prototypes)
 * - `Uint8Array`
 * - `Signal` from `@preact/signals`
 *
 * Circular references are supported and objects with the same reference are
 * serialized only once.
 *
 * The corresponding deserializer is in `src/runtime/deserializer.ts`.
 */
import { isValidElement, VNode } from "preact";
import { KEY } from "../runtime/deserializer.ts";

interface SerializeResult {
  /** The string serialization. */
  serialized: string;
  /** If the deserializer is required to deserialize this string. If this is
   * `false` the serialized string can be deserialized with `JSON.parse`. */
  requiresDeserializer: boolean;
  /** If the serialization contains serialized signals. If this is `true` the
   * deserializer must be passed a factory functions for signals. */
  hasSignals: boolean;
}

interface Signal {
  peek(): unknown;
  value: unknown;
}

// deno-lint-ignore no-explicit-any
function isSignal(x: any): x is Signal {
  return (
    x !== null &&
    typeof x === "object" &&
    typeof x.peek === "function" &&
    "value" in x
  );
}

// deno-lint-ignore no-explicit-any
function isVNode(x: any): x is VNode {
  return x !== null && typeof x === "object" && "type" in x && "ref" in x &&
    "__k" in x &&
    isValidElement(x);
}

export function serialize(data: unknown): SerializeResult {
  let requiresDeserializer = false;
  let hasSignals = false;
  const seen = new Map<unknown, (string | null)[]>();
  const references = new Map<(string | null)[], (string | null)[][]>();

  const keyStack: (string | null)[] = [];
  const parentStack: unknown[] = [];

  let earlyReturn = false;

  const toSerialize = {
    v: data,
    get r() {
      earlyReturn = true;
      if (references.size > 0) {
        const refs = [];
        for (const [targetPath, refPaths] of references) {
          refs.push([targetPath, ...refPaths]);
        }
        return refs;
      }
      return undefined;
    },
  };

  function replacer(
    this: unknown,
    key: string | null,
    value: unknown,
  ): unknown {
    if (value === toSerialize || earlyReturn) {
      return value;
    }

    // Bypass signal's `.toJSON` method because we want to serialize
    // the signal itself including the signal's value and not just
    // the value. This is needed because `JSON.stringify` always
    // calls `.toJSON` automatically if available.
    // deno-lint-ignore no-explicit-any
    if (key !== null && isSignal((this as any)[key])) {
      // deno-lint-ignore no-explicit-any
      value = (this as any)[key];
    }

    // For some object types, the path in the object graph from root is not the
    // same between the serialized representation, and deserialized objects. For
    // these cases, we have to change the contents of the key stack to match the
    // deserialized object.
    if (typeof this === "object" && this !== null && KEY in this) {
      if (this[KEY] === "s" && key === "v") key = "value"; // signals
      if (this[KEY] === "l" && key === "v") key = null; // literals (magic key object)
    }

    if (this !== toSerialize) {
      const parentIndex = parentStack.indexOf(this);
      parentStack.splice(parentIndex + 1);
      keyStack.splice(parentIndex);
      keyStack.push(key);
      // the parent is pushed before return
    }

    if (typeof value === "object" && value !== null) {
      const path = seen.get(value);
      const currentPath = [...keyStack];
      if (path !== undefined) {
        requiresDeserializer = true;
        const referenceArr = references.get(path);
        if (referenceArr === undefined) {
          references.set(path, [currentPath]);
        } else {
          referenceArr.push(currentPath);
        }
        return 0;
      } else if (isVNode(value)) {
        requiresDeserializer = true;
        // No need to serialize JSX as we pick that up from
        // the rendered HTML in the browser.
        const res = null;
        parentStack.push(res);
        return res;
      } else {
        seen.set(value, currentPath);
      }
    }

    if (isSignal(value)) {
      requiresDeserializer = true;
      hasSignals = true;
      const res = { [KEY]: "s", v: value.peek() };
      parentStack.push(res);
      return res;
    } else if (typeof value === "bigint") {
      requiresDeserializer = true;
      const res = { [KEY]: "b", d: value.toString() };
      parentStack.push(res);
      return res;
    } else if (value instanceof Uint8Array) {
      requiresDeserializer = true;
      const res = { [KEY]: "u8a", d: b64encode(value) };
      parentStack.push(res);
      return res;
    } else if (typeof value === "object" && value && KEY in value) {
      requiresDeserializer = true;
      // deno-lint-ignore no-explicit-any
      const v: any = { ...value };
      const k = v[KEY];
      delete v[KEY];
      const res = { [KEY]: "l", k, v };
      parentStack.push(res);
      return res;
    } else {
      parentStack.push(value);
      return value;
    }
  }

  const serialized = JSON.stringify(toSerialize, replacer);
  return { serialized, requiresDeserializer, hasSignals };
}

// deno-fmt-ignore
const base64abc = [
  "A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O",
  "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z", "a", "b", "c", "d",
  "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s",
  "t", "u", "v", "w", "x", "y", "z", "0", "1", "2", "3", "4", "5", "6", "7",
  "8", "9", "+", "/",
];

/**
 * CREDIT: https://gist.github.com/enepomnyaschih/72c423f727d395eeaa09697058238727
 * Encodes a given Uint8Array, ArrayBuffer or string into RFC4648 base64 representation
 */
export function b64encode(buffer: ArrayBuffer): string {
  const uint8 = new Uint8Array(buffer);
  let result = "",
    i;
  const l = uint8.length;
  for (i = 2; i < l; i += 3) {
    result += base64abc[uint8[i - 2] >> 2];
    result += base64abc[((uint8[i - 2] & 0x03) << 4) | (uint8[i - 1] >> 4)];
    result += base64abc[((uint8[i - 1] & 0x0f) << 2) | (uint8[i] >> 6)];
    result += base64abc[uint8[i] & 0x3f];
  }
  if (i === l + 1) {
    // 1 octet yet to write
    result += base64abc[uint8[i - 2] >> 2];
    result += base64abc[(uint8[i - 2] & 0x03) << 4];
    result += "==";
  }
  if (i === l) {
    // 2 octets yet to write
    result += base64abc[uint8[i - 2] >> 2];
    result += base64abc[((uint8[i - 2] & 0x03) << 4) | (uint8[i - 1] >> 4)];
    result += base64abc[(uint8[i - 1] & 0x0f) << 2];
    result += "=";
  }
  return result;
}
