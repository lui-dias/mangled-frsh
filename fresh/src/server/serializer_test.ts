// deno-lint-ignore-file no-explicit-any

import { serialize } from "./serializer.ts";
import { assert, assertEquals, assertSnapshot } from "../../tests/deps.ts";
import { deserialize, KEY } from "../runtime/deserializer.ts";
import { signal } from "@preact/signals-core";
import { signal as signal130 } from "@preact/signals-core@1.3.0";
import { signal as signal123 } from "@preact/signals-core@1.2.3";

Deno.test("serializer - primitives & plain objects", async (t) => {
  const data = {
    a: 1,
    b: "2",
    c: true,
    d: null,
    f: [1, 2, 3],
    g: { a: 1, b: 2, c: 3 },
  };
  const res = serialize(data);
  assert(!res.requiresDeserializer);
  assert(!res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - bigint", async (t) => {
  const data = { a: 999999999999999999n };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(!res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - Uint8Array", async (t) => {
  const data = { a: new Uint8Array([1, 2, 3]) };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(!res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - signals", async (t) => {
  const data = {
    a: 1,
    b: signal(2),
  };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized, signal);
  assertEquals(typeof deserialized, "object");
  assertEquals(deserialized.a, 1);
  assertEquals(deserialized.b.value, 2);
  assertEquals(deserialized.b.peek(), 2);
});

Deno.test("serializer - @preact/signals-core 1.2.3", async (t) => {
  const data = {
    a: 1,
    b: signal123(2),
  };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized, signal);
  assertEquals(typeof deserialized, "object");
  assertEquals(deserialized.a, 1);
  assertEquals(deserialized.b.value, 2);
  assertEquals(deserialized.b.peek(), 2);
});

Deno.test("serializer - @preact/signals-core 1.3.0", async (t) => {
  const data = {
    a: 1,
    b: signal130(2),
  };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized, signal);
  assertEquals(typeof deserialized, "object");
  assertEquals(deserialized.a, 1);
  assertEquals(deserialized.b.value, 2);
  assertEquals(deserialized.b.peek(), 2);
});

Deno.test("serializer - multiple versions of @preact/signals-core", async (t) => {
  const data = {
    a: 1,
    b: signal(2),
    c: signal123(2),
    d: signal130(2),
  };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized, signal);
  assertEquals(typeof deserialized, "object");
  assertEquals(deserialized.a, 1);
  assertEquals(deserialized.b.value, 2);
  assertEquals(deserialized.b.peek(), 2);
  assertEquals(deserialized.c.value, 2);
  assertEquals(deserialized.c.peek(), 2);
  assertEquals(deserialized.d.value, 2);
  assertEquals(deserialized.d.peek(), 2);
});

Deno.test("serializer - magic key", async (t) => {
  const data = { [KEY]: "f", a: 1 };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(!res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - circular reference objects", async (t) => {
  const data: any = { a: 1 };
  data.b = data;
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(!res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - circular reference nested objects", async (t) => {
  const data: any = { a: 1, b: { c: 2 } };
  data.b.d = data;
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(!res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - circular reference array", async (t) => {
  const data: any = [1, 2, 3];
  data.push(data);
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(!res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized);
  assertEquals(deserialized, data);
  assertEquals(deserialized.length, 4);
});

Deno.test("serializer - multiple reference", async (t) => {
  const data: any = { a: 1, b: { c: 2 } };
  data.d = data.b;
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(!res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - multiple reference signals", async (t) => {
  const inner: any = { [KEY]: "x", x: 1 };
  inner.y = inner;
  const s = signal(inner);
  const data = { inner, a: s, b: { c: s } };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized, signal);
  assertEquals(deserialized.a.value, inner);
  assertEquals(deserialized.a.peek(), inner);
  assertEquals(deserialized.b.c.value, inner);
  assertEquals(deserialized.b.c.peek(), inner);
  deserialized.a.value = 2;
  assertEquals(deserialized.a.value, 2);
  assertEquals(deserialized.b.c.value, 2);
});

Deno.test("serializer - multiple reference in magic key", async (t) => {
  const inner = { foo: "bar" };
  const literal: any = { [KEY]: "x", inner };
  const data = { literal, inner };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(!res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized);
  assertEquals(deserialized, data);
});

Deno.test("serializer - multiple reference in signal", async (t) => {
  const inner = { foo: "bar" };
  const s = signal(inner);
  const data = { s, inner };
  const res = serialize(data);
  assert(res.requiresDeserializer);
  assert(res.hasSignals);
  await assertSnapshot(t, res.serialized);
  const deserialized: any = deserialize(res.serialized, signal);
  assertEquals(deserialized.s.value, inner);
  assertEquals(deserialized.s.peek(), inner);
  assertEquals(deserialized.inner, inner);
});
