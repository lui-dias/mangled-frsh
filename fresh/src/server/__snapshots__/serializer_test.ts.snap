export const snapshot = {};

snapshot[`serializer - primitives & plain objects 1`] = `'{"v":{"a":1,"b":"2","c":true,"d":null,"f":[1,2,3],"g":{"a":1,"b":2,"c":3}}}'`;

snapshot[`serializer - bigint 1`] = `'{"v":{"a":{"_f":"b","d":"999999999999999999"}}}'`;

snapshot[`serializer - Uint8Array 1`] = `'{"v":{"a":{"_f":"u8a","d":"AQID"}}}'`;

snapshot[`serializer - signals 1`] = `'{"v":{"a":1,"b":{"_f":"s","v":2}}}'`;

snapshot[`serializer - @preact/signals-core 1.2.3 1`] = `'{"v":{"a":1,"b":{"_f":"s","v":2}}}'`;

snapshot[`serializer - @preact/signals-core 1.3.0 1`] = `'{"v":{"a":1,"b":{"_f":"s","v":2}}}'`;

snapshot[`serializer - multiple versions of @preact/signals-core 1`] = `'{"v":{"a":1,"b":{"_f":"s","v":2},"c":{"_f":"s","v":2},"d":{"_f":"s","v":2}}}'`;

snapshot[`serializer - magic key 1`] = `'{"v":{"_f":"l","k":"f","v":{"a":1}}}'`;

snapshot[`serializer - circular reference objects 1`] = `'{"v":{"a":1,"b":0},"r":[[[],["b"]]]}'`;

snapshot[`serializer - circular reference nested objects 1`] = `'{"v":{"a":1,"b":{"c":2,"d":0}},"r":[[[],["b","d"]]]}'`;

snapshot[`serializer - circular reference array 1`] = `'{"v":[1,2,3,0],"r":[[[],["3"]]]}'`;

snapshot[`serializer - multiple reference 1`] = `'{"v":{"a":1,"b":{"c":2},"d":0},"r":[[["b"],["d"]]]}'`;

snapshot[`serializer - multiple reference signals 1`] = `'{"v":{"inner":{"_f":"l","k":"x","v":{"x":1,"y":0}},"a":{"_f":"s","v":0},"b":{"c":0}},"r":[[["inner"],["inner",null,"y"],["a","value"]],[["a"],["b","c"]]]}'`;

snapshot[`serializer - multiple reference in magic key 1`] = `'{"v":{"literal":{"_f":"l","k":"x","v":{"inner":{"foo":"bar"}}},"inner":0},"r":[[["literal",null,"inner"],["inner"]]]}'`;

snapshot[`serializer - multiple reference in signal 1`] = `'{"v":{"s":{"_f":"s","v":{"foo":"bar"}},"inner":0},"r":[[["s","value"],["inner"]]]}'`;
